crates/wascc-provider/src/states/container/waiting.rs

```diff
@@ -0,0 +1,272 @@
+    let mut lock = port_map.lock().await;
+    while empty_port.len() < 2768 {
+        let generated_port: u16 = rand::thread_rng().gen_range(30000, 32768);
+        if !lock.contains_key(&generated_port) {
+            lock.insert(generated_port, pod_key);
+            return Ok(generated_port);
+        }
+        empty_port.insert(generated_port);
+    }
+    Err(PortAllocationError)
+}
+
+async fn assign_container_port(
+    port_map: Arc<Mutex<BTreeMap<u16, PodKey>>>,
+    pod: &Pod,
+    container: &Container,
+) -> anyhow::Result<u16> {
+    let mut port_assigned: u16 = 0;
+    if let Some(container_vec) = container.ports().as_ref() {
+        for c_port in container_vec.iter() {
+            let container_port = c_port.container_port;
+            if let Some(host_port) = c_port.host_port {
+                let host_port: u16 = u16::try_from(host_port)?;
+                let mut lock = port_map.lock().await;
+                if !lock.contains_key(&host_port) {
+                    port_assigned = host_port;
+                    lock.insert(port_assigned, PodKey::from(pod));
+                } else {
+                    error!(
+                        "Failed to assign hostport {}, because it's taken",
+                        &host_port
+                    );
+                    return Err(anyhow::anyhow!("Port {} is currently in use", &host_port));
+                }
+            } else if container_port >= 0 && container_port <= 65536 {
+                port_assigned = find_available_port(&port_map, pod).await?;
+            }
+        }
+    }
+    Ok(port_assigned)
+}
+
+/// The container is starting.
+#[derive(Default, Debug, TransitionTo)]
+#[transition_to(Running, Terminated)]
+pub struct Waiting;
+
+#[async_trait::async_trait]
+impl State<ContainerState> for Waiting {
+    async fn next(
+        self: Box<Self>,
+        shared: SharedState<ProviderState>,
+        state: &mut ContainerState,
+        container: &Container,
+    ) -> Transition<ContainerState> {
+        info!(
+            "Starting container {} for pod {}",
+            container.name(),
+            state.pod.name(),
+        );
+
+        let port_assigned = {
+            let port_map = shared.read().await.port_map.clone();
+            match assign_container_port(Arc::clone(&port_map), &state.pod, &container).await {
+                Ok(port) => port,
+                Err(e) => {
+                    return Transition::next(
+                        self,
+                        Terminated::new(
+                            format!(
+                                "Pod {} container {} failed to allocate port: {:?}",
+                                state.pod.name(),
+                                container.name(),
+                                e
+                            ),
+                            true,
+                        ),
+                    )
+                }
+            }
+        };
```

Comment:
~I find this much nesting really hard to parse.  Can we simplify this in any way?~  Looking at the broader context I now suspect this is existing code and we are just moving it into a state with the minimum possible change and risk, so can probably ignore this for now.

Reply:
I definitely think these things should be broken up in the future.